server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java

```diff
@@ -0,0 +1,141 @@
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateTaskExecutor;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.desirednodes.DesiredNodesSettingsValidator;
+import org.elasticsearch.cluster.desirednodes.VersionConflictException;
+import org.elasticsearch.cluster.metadata.DesiredNodes;
+import org.elasticsearch.cluster.metadata.DesiredNodesMetadata;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.Locale;
+
+import static java.lang.String.format;
+
+public class TransportUpdateDesiredNodesAction extends TransportMasterNodeAction<UpdateDesiredNodesRequest, UpdateDesiredNodesResponse> {
+    private final DesiredNodesSettingsValidator settingsValidator;
+
+    @Inject
+    public TransportUpdateDesiredNodesAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        DesiredNodesSettingsValidator settingsValidator
+    ) {
+        super(
+            UpdateDesiredNodesAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            UpdateDesiredNodesRequest::new,
+            indexNameExpressionResolver,
+            UpdateDesiredNodesResponse::new,
+            ThreadPool.Names.SAME
+        );
+        this.settingsValidator = settingsValidator;
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(UpdateDesiredNodesRequest request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        UpdateDesiredNodesRequest request,
+        ClusterState state,
+        ActionListener<UpdateDesiredNodesResponse> listener
+    ) throws Exception {
+        try {
+            DesiredNodes proposedDesiredNodes = new DesiredNodes(request.getHistoryID(), request.getVersion(), request.getNodes());
+            settingsValidator.validate(proposedDesiredNodes);
+
+            clusterService.submitStateUpdateTask(
+                "update-desired-nodes",
+                new ClusterStateUpdateTask(Priority.HIGH, request.masterNodeTimeout()) {
```

Comment:
I would think we should use `URGENT` here instead, but maybe @DaveCTurner has an opinion on that?

Reply:
I'd be ok with `URGENT` although I'd be happier about it if these things were batched, just in case the orchestrator goes haywire.

Reply:
I'll update the PR batching them